src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala

```diff
@@ -50,60 +51,102 @@ class TaskTracker @Inject() (
     }
   }
 
-  def terminated(appId: PathId, taskId: String): Future[Option[MarathonTask]] = {
-    val appTasks = getInternal(appId)
-    val app = apps(appId)
+  private[this] def processUpdate(app: InternalApp, task: MarathonTask, status: TaskStatus): Future[_] = {
+    def updateTaskOnStateChange(task: MarathonTask): Future[_] = {
+      def statusDidChange(statusA: TaskStatus, statusB: TaskStatus): Boolean = {
+        val healthy = statusB.hasHealthy &&
+          (!statusA.hasHealthy || statusA.getHealthy != statusB.getHealthy)
 
-    appTasks.get(taskId) match {
-      case Some(task) =>
-        app.tasks.remove(task.getId)
+        healthy || statusA.getState != statusB.getState
+      }
 
-        Await.result(repo.expunge(taskId), timeout)
+      if (statusDidChange(task.getStatus, status)) {
+        updateTask(task.toBuilder.setStatus(status).build)
+      }
+      else {
+        log.debug(s"Ignoring status update for ${task.getId}. Status did not change.")
+        Future.successful(())
+      }
+    }
 
-        log.info(s"Task $taskId expunged and removed from TaskTracker")
+    def updateTask(updatedTask: MarathonTask): Future[_] = {
+      val appId = app.appName
+      val taskId: String = updatedTask.getId
+      log.debug(s"updating task [$taskId] of app [$appId]: {}", updatedTask)
 
-        if (app.shutdown && app.tasks.isEmpty) {
-          // Are we shutting down this app? If so, remove it
-          remove(appId)
-        }
+      getInternal(appId) += (taskId -> updatedTask)
+      repo.store(updatedTask).map(_ => updatedTask)
+    }
 
-        Future.successful(Some(task))
-      case None =>
-        if (app.shutdown && app.tasks.isEmpty) {
-          // Are we shutting down this app? If so, remove it
-          remove(appId)
-        }
-        Future.successful(None)
+    status.getState match {
+      case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
+        removeTask(app, task.getId)
+        Future.successful(())
+      case TASK_RUNNING if !task.hasStartedAt => // was staged, is now running
+        updateTask(task.toBuilder.setStartedAt(System.currentTimeMillis).setStatus(status).build())
+      case _ =>
+        updateTaskOnStateChange(task)
+    }
+  }
+
+  /**
+    * Remove the task for the given app with the given ID completely.
+    *
+    * If the task does not exist, the method will not fail.
+    */
+  def terminated(appId: PathId, taskId: String): Future[_] = {
+    val maybeInternalApp = apps.get(appId)
+    maybeInternalApp.foreach(app => removeTask(app, taskId))
+    Future.successful(())
+  }
+
+  private[this] def removeTask(app: InternalApp, taskId: String): Unit = {
+    app.tasks.remove(taskId)
+
+    Await.result(repo.expunge(taskId), timeout)
+
+    log.info(s"Task $taskId expunged and removed from TaskTracker")
+
+    if (app.shutdown && app.tasks.isEmpty) {
+      // Are we shutting down this app? If so, remove it
+      remove(app.appName)
     }
   }
 
```

Comment:
Aaaah, @aquamatthias you made the comment above the method. github does not display the relevant code snippet then. Now I get you. `shutdown` should not use `getOrElseUpdate`. But since we did not touch this code, I typically would't want to change this code in this PR. But I will...